R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
+#'   xValues = c(0, 14.482, 28.965, 0, 1, 2),
+#'   xUnit = "min", xDimension = "Time",
+#'   yValues = c(1, 1, 1, 1, 1, 1),
+#'   yUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   yDimension = c("Amount", "Amount", "Amount", "Mass", "Mass", "Mass"),
+#'   yErrorValues = c(2.747, 2.918, 2.746, NA, NA, NA),
+#'   yErrorUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   molWeight = c(10, 10, 20, 20, 20, 10)
+#' ))
+#'
+#' # default conversion
+#' ospsuite:::.unitConverter(df)
+#'
+#' # customizing conversion with specified unit(s)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Mass$kg)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$s, yUnit = ospUnits$Amount$mmol)
+#'
+#' @keywords internal
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # target units --------------------------
+
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have the same units for
+  # visual/graphical comparison.
+  #
+  # Therefore, if target units are not specified by the user, we need to choose
+  # one ourselves. For no special reason, the first element from a vector of
+  # unique units will be selected: one for X-axis, and one for Y-axis, i.e.
+  xTargetUnit <- xUnit %||% unique(data$xUnit)[[1]]
+  yTargetUnit <- yUnit %||% unique(data$yUnit)[[1]]
+
+  # internal --------------------------
+
+  # The strategy is to split the data frame for each source unit and carry out
+  # conversion separately per data frame. This is the most performant options
+  # since there can only be as many expensive calls to `toUnit()`/`rClr` as
+  # there are source units.
+  #
+  # The problem occurs when source units are missing (`NA`). The `toUnit`
+  # function can handle them but not `split()`, which would drop the entire
+  # section of the data frame corresponding to `NA` and thus there will be loss
+  # of data when a data frame is split into a list of data frames.
+  #
+  # The trick is to create copies of source unit and molecular weight columns
+  # and fill in the missing values with something other than `NA`. Using these
+  # new columns with `split()` makes sure that parts of a data frame where
+  # source units are missing won't be dropped. Note that the original columns
+  # containing source units remain untouched.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
+        .fns = as.character,
+        .names = "{.col}Split"
+      )
+    ) %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Split$"),
+        .fns = ~ tidyr::replace_na(.x, "missing")
+      )
+    ) %>%
+    # `split()` will change the row order of the data frame depending on the
```

Comment:
you talk about split() everywhere but I don't see it being used. Am i blind?

Reply:
<img width="798" alt="Screenshot 2022-04-22 at 17 45 02" src="https://user-images.githubusercontent.com/11330453/164749183-ddc4faba-ff12-4757-8487-6d6193cf3a6f.png">

Reply:
Ok so you talk about split() on line 424 but the usage is online 458. (more than 30 lines below). 
Why don't you comment close to where something is happening? Mentally, it is very hard to read all those comments if they are not directly related to the code. (at least for me).

Reply:
> Ok so you talk about split() on line 424 but the usage is online 458

Because I _need to_. Otherwise, you would be wondering, 

> "why the heck are you doing all this?! Why are you adding _Split_ suffixes? Why are you replacing missing values with 'missing'? 

I need to set the stage for what's coming. If I mention all this when `split()` is used for the first time, it's already too late. Reader is already frustrated by not having answers to all the questions above.